File: Swift/LeetCode/LeetCode/MergeIntervals.swift
//
//  MergeIntervals.swift
//  LeetCode
//
//  Created by 黎赵太郎 on 07/01/2018.
//  Copyright © 2018 lizhaotailang. All rights reserved.
//

import Foundation

class MergeIntervals {
    
    func merge(_ intervals: [Interval]) -> [Interval] {
        if intervals.isEmpty || intervals.count == 1 {
            return intervals
        }
        
        var results = Array<Interval>.init()
        var newIntervals = intervals.sorted{ $0.start < $1.start}
        
        results.append(newIntervals[0])
        var i = 1
        while i < newIntervals.count {
            if results[results.count - 1].end >= newIntervals[i].start {
                results[results.count - 1].end = max(results[results.count - 1].end, newIntervals[i].end)
            } else {
                results.append(newIntervals[i])
            }
            
            i += 1
        }
        
        return results
    }
    
    public class Interval: CustomStringConvertible {
        public var start: Int
        public var end: Int
        public init(_ start: Int, _ end: Int) {
            self.start = start
            self.end = end
        }
        
        public var description: String {
            return "Interval start: \(self.start) end: -> \(self.end)) "
        }
    }
    
}

File: Golang/MergeIntervals/MergeIntervals.go
package MergeIntervals

import (
	"sort"
)

//Given a collection of intervals, merge all overlapping intervals.
//
//For example,
//Given [1,3],[2,6],[8,10],[15,18],
//return [1,6],[8,10],[15,18].
//
//Accepted.

type Interval struct {
	Start int
	End   int
}

func merge(intervals []Interval) []Interval {
	if intervals == nil || len(intervals) == 0 || len(intervals) == 1 {
		return intervals
	}

	var results []Interval

	sort.Slice(intervals, func(i, j int) bool {
		return intervals[i].Start < intervals[j].Start
	})

	results = append(results, intervals[0])
	intervalsSize := len(intervals)
	for i := 1; i < intervalsSize; i++ {
		if results[len(results)-1].End >= intervals[i].Start {
			results[len(results)-1].End = max(results[len(results)-1].End, intervals[i].End)
		} else {
			results = append(results, intervals[i])
		}
	}

	return results
}

func max(i, j int) int {
	if i > j {
		return i
	}
	return j
}
